core/src/main/java/org/apache/iceberg/types/Serializers.java

```diff
@@ -0,0 +1,237 @@
+  private static <T> Serializer<T> internal(Type type) {
+    if (type.isPrimitiveType()) {
+      return forType(type.asPrimitiveType());
+    } else if (type.isStructType()) {
+      return (Serializer<T>) forType(type.asStructType());
+    } else if (type.isListType()) {
+      return (Serializer<T>) forType(type.asListType());
+    }
+    throw new UnsupportedOperationException("Cannot determine serializer for type: " + type);
+  }
+
+  @SuppressWarnings("unchecked")
+  private static <T> Class<T> internalClass(Type type) {
+    if (type.isPrimitiveType()) {
+      return (Class<T>) type.typeId().javaClass();
+    } else if (type.isStructType()) {
+      return (Class<T>) StructLike.class;
+    } else if (type.isListType()) {
+      return (Class<T>) List.class;
+    } else if (type.isMapType()) {
+      return (Class<T>) Map.class;
+    }
+
+    throw new UnsupportedOperationException("Cannot determine expected class for type: " + type);
+  }
+
+  private static class StructLikeSerializer implements Serializer<StructLike> {
+    private final Types.StructType struct;
+    private final Serializer<Object>[] serializers;
+    private final Class<?>[] classes;
+
+    private StructLikeSerializer(Types.StructType struct) {
+      this.struct = struct;
+      this.serializers = struct.fields().stream()
+          .map(field -> internal(field.type()))
+          .toArray((IntFunction<Serializer<Object>[]>) Serializer[]::new);
+      this.classes = struct.fields().stream()
+          .map(field -> internalClass(field.type()))
+          .toArray(Class<?>[]::new);
+    }
+
+    @Override
+    public byte[] serialize(StructLike object) {
+      if (object == null) {
+        return null;
+      }
+
+      try (ByteArrayOutputStream out = new ByteArrayOutputStream();
+           DataOutputStream dos = new DataOutputStream(out)) {
+
+        for (int i = 0; i < serializers.length; i += 1) {
+          Class<?> valueClass = classes[i];
+
+          byte[] fieldData = serializers[i].serialize(object.get(i, valueClass));
+          if (fieldData == null) {
+            dos.writeInt(NULL_MARK);
+          } else {
+            dos.writeInt(fieldData.length);
+            dos.write(fieldData);
+          }
+        }
+        return out.toByteArray();
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+    }
+
+    @Override
+    public StructLike deserialize(byte[] data) {
+      if (data == null) {
+        return null;
+      }
+
+      try (ByteArrayInputStream in = new ByteArrayInputStream(data);
+           DataInputStream dis = new DataInputStream(in)) {
+
+        GenericRecord record = GenericRecord.create(struct);
+        for (int i = 0; i < serializers.length; i += 1) {
+          int length = dis.readInt();
+
+          if (length == NULL_MARK) {
+            record.set(i, null);
+          } else {
+            byte[] fieldData = new byte[length];
+            int fieldDataSize = dis.read(fieldData);
```

Comment:
I think there should be `int fieldDataSize = length > 0 ? dis.read(fieldData) : 0;`.
Because `dis.read()` will return -1 when fieldData lenght is 0(e.g. empty string will cause this) and make `Preconditions.checkState` fail.